File: dll/argtypes/dll/argtypes_list.c
#include <stdio.h>

void at_lst_1d_int(int* l, int len)
{
    printf("at_lst_1d_int: ");
    for (int i = 0; i < len; i++)
    {
        printf("%d ", l[i]);
    }
    printf("\n");
}

void at_lst_1d_dbl(double* l, int len)
{
    printf("at_lst_1d_dbl: ");
    for (int i = 0; i < len; i++)
    {
        printf("%.1f ", l[i]);
    }
    printf("\n");
}

void at_lst_2d_int(int** l, int rows, int cols)
{
    printf("at_lst_2d_int:\n");
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            printf("%d ", l[i][j]);
        }
        printf("\n");
    }
}

void at_lst_2d_dbl(double** l, int rows, int cols)
{
    printf("at_lst_2d_double:\n");
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            printf("%.1f ", l[i][j]);
        }
        printf("\n");
    }
}

File: dll/argtypes/argtypes_list.py
import capi
from random import randint

dll = "dll/argtypes.dll"


@capi.func(dll)
def at_lst_1d_int(l, i):
    pass


@capi.func(dll)
def at_lst_1d_dbl(l, i):
    pass


@capi.func(dll)
def at_lst_2d_int(l, r, c):
    pass


@capi.func(dll)
def at_lst_2d_dbl(l, r, c):
    pass


n = 10
il = [[randint(100, 999) for i in range(n)] for j in range(n)]
fl = [[float(e) for e in r] for r in il]

at_lst_1d_int(il[0], n)
at_lst_1d_dbl(fl[0], n)

at_lst_2d_int(il, n, n)
at_lst_2d_dbl(fl, n, n)
